File: sdk/go/tools.go
package llmengine

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

type ToolsClient struct {
	BaseURL    string
	HTTPClient *http.Client
	APIKey     string
}

func NewToolsClient(baseURL string, apiKey string) *ToolsClient {
	return &ToolsClient{
		BaseURL:    baseURL,
		HTTPClient: &http.Client{},
		APIKey:     apiKey,
	}
}

type ToolMetadata struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	Category    string  `json:"category"`
	Cost        float64 `json:"cost_estimate"`
}

type ToolResult struct {
	ExecutionID   string      `json:"execution_id"`
	ToolName      string      `json:"tool_name"`
	Status        string      `json:"status"`
	Output        interface{} `json:"output"`
	Error         string      `json:"error"`
	ExecutionTime float64     `json:"execution_time_seconds"`
}

func (c *ToolsClient) Create(name string, description string) (map[string]interface{}, error) {
	url := fmt.Sprintf("%s/v1/tools", c.BaseURL)
	payload := map[string]interface{}{
		"name":        name,
		"description": description,
	}
	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(body))
	req.Header.Set("Content-Type", "application/json")
	if c.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.APIKey)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	return result, nil
}

func (c *ToolsClient) List(tenantID string) ([]ToolMetadata, error) {
	url := fmt.Sprintf("%s/v1/tools?tenant_id=%s", c.BaseURL, tenantID)
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var res map[string][]ToolMetadata
	json.NewDecoder(resp.Body).Decode(&res)
	return res["tools"], nil
}

func (c *ToolsClient) Execute(toolID string, params map[string]interface{}) (*ToolResult, error) {
	url := fmt.Sprintf("%s/v1/tools/%s/execute", c.BaseURL, toolID)
	payload := map[string]interface{}{
		"parameters": params,
	}
	body, _ := json.Marshal(payload)
	resp, err := http.Post(url, "application/json", bytes.NewBuffer(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var res map[string]ToolResult
	json.NewDecoder(resp.Body).Decode(&res)
	r := res["result"]
	return &r, nil
}

File: sdk/typescript/src/tools.ts
/**
 * TypeScript SDK Client for Enterprise Tool Calling & MCP Platform
 */

export interface ToolMetadata {
  name: string;
  description: string;
  category: string;
  cost_estimate?: number;
  requires_approval?: boolean;
}

export interface ToolResult {
  execution_id: string;
  tool_name: string;
  status: string;
  output?: any;
  error?: string;
  execution_time_seconds: number;
  cost: number;
}

export class ToolsClient {
  private baseUrl: string;
  private apiKey?: string;

  constructor(baseUrl: string = "http://localhost:8000", apiKey?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }
    return headers;
  }

  async create(tool: Partial<ToolMetadata>): Promise<{ status: string; tool: ToolMetadata }> {
    const res = await fetch(`${this.baseUrl}/v1/tools`, {
      method: "POST",
      headers: this.getHeaders(),
      body: JSON.stringify(tool),
    });
    return res.json();
  }

  async list(tenantId: string = "global"): Promise<ToolMetadata[]> {
    const res = await fetch(`${this.baseUrl}/v1/tools?tenant_id=${tenantId}`, {
      headers: this.getHeaders(),
    });
    const data = await res.json();
    return data.tools || [];
  }

  async get(toolId: string, tenantId: string = "global"): Promise<ToolMetadata> {
    const res = await fetch(`${this.baseUrl}/v1/tools/${toolId}?tenant_id=${tenantId}`, {
      headers: this.getHeaders(),
    });
    const data = await res.json();
    return data.tool;
  }

  async delete(toolId: string, tenantId: string = "global"): Promise<{ status: string; id: string }> {
    const res = await fetch(`${this.baseUrl}/v1/tools/${toolId}?tenant_id=${tenantId}`, {
      method: "DELETE",
      headers: this.getHeaders(),
    });
    return res.json();
  }

  async execute(toolId: string, parameters: Record<string, any>, context?: Record<string, any>): Promise<ToolResult> {
    const res = await fetch(`${this.baseUrl}/v1/tools/${toolId}/execute`, {
      method: "POST",
      headers: this.getHeaders(),
      body: JSON.stringify({ parameters, context }),
    });
    const data = await res.json();
    return data.result;
  }

  async validate(toolId: string, parameters: Record<string, any>): Promise<{ valid: boolean; error?: string }> {
    const res = await fetch(`${this.baseUrl}/v1/tools/${toolId}/validate`, {
      method: "POST",
      headers: this.getHeaders(),
      body: JSON.stringify({ parameters }),
    });
    return res.json();
  }

  async audit(toolId: string, tenantId: string = "global"): Promise<{ audit_logs: any[] }> {
    const res = await fetch(`${this.baseUrl}/v1/tools/${toolId}/audit?tenant_id=${tenantId}`, {
      headers: this.getHeaders(),
    });
    return res.json();
  }

  async metrics(toolId: string, tenantId: string = "global"): Promise<{ metrics: any }> {
    const res = await fetch(`${this.baseUrl}/v1/tools/${toolId}/metrics?tenant_id=${tenantId}`, {
      headers: this.getHeaders(),
    });
    return res.json();
  }
}
